manual/manual/html_processing/src/process_api.ml

```diff
@@ -0,0 +1,380 @@
+
+open Soup
+open Printf
+open Common
+
+let compiler_libref = ref false
+(* set this to true to process compilerlibref instead of libref *)
+
+type config = {
+  src_dir : string;
+  dst_dir : string;
+  title : string
+}
+
+let favicon = "favicon.ico"
+
+(* HTML code for the search widget. We don't add the "onchange" event because it
+   forces to click twice to an external link after entering text. *)
+let search_widget with_description =
+  let search_decription = if with_description
+    then {|<span class="search_comment">(search values, type signatures, and descriptions - case sensitive)<div class="search_help"><ul><li>You may search bare values, like <code>map</code>, or indicate the module, like <code>List.map</code>, or type signatures, like <code>int -> float</code>.</li><li>To combine several keywords, just separate them by a space. Quotes "like this" can be used to prevent from splitting words at spaces.</li><li>You may use the special chars <code>^</code> and <code>$</code> to indicate where the matched string should start or end, respectively.</li></ul></div></span>|}
+    else "" in
+  sprintf {|<div class="api_search"><input type="text" name="apisearch" id="api_search" class="api_search"
+	 oninput    = "mySearch(%b);"
+         onkeypress = "this.oninput();"
+         onclick    = "this.oninput();"
+	 onpaste    = "this.oninput();">
+<img src="search_icon.svg" alt="Search" class="api_search svg" onclick="mySearch(%b)">%s</div>
+<div id="search_results"></div>|} with_description with_description search_decription
+  |> parse
+
+(* We save parsed files in a table; this is just for speed optimization,
+   especially for make_index (18sec instead of 50sec for the whole index); it
+   can be removed.  Although if we really wanted a fast make_index, we would use
+   Scanf all over the place ==> 1sec. Warning: the parsed files will be mutated
+   by processing, so one should never process the same file twice. *)
+
+let parsed_files = Hashtbl.create 50
+
+let parse_file ?(original=false) file =
+  match Hashtbl.find_opt parsed_files file with
+  | Some soup ->
+      if original then failwith (sprintf "File %s was already processed" file)
+      else soup
+  | None ->
+      let soup = read_file file |> parse in
+      Hashtbl.add parsed_files file soup;
+      soup
+
+let process ?(search=true) ~version config file out =
+
+  dbg "Processing %s..." file;
+  let soup = parse_file ~original:true file in
+
+  (* Add javascript files *)
+  let head = soup $ "head" in
+  if search then begin
+    create_element "script" ~attributes:["src","search.js"]
+    |> append_child head
+  end;
+  create_element "script" ~attributes:["src","scroll.js"]
+  |> append_child head;
+  create_element "script" ~attributes:["src","navigation.js"]
+  |> append_child head;
+
+  (* Add favicon *)
+  let head = soup $ "head" in
+  parse ({|<link rel="shortcut icon" type="image/x-icon" href="|} ^ favicon ^ {|">|})
+  |> append_child head;
```

Comment:
The separation and top comments are a good idea. I would propose to go one step further and make all those parts of the processing pipeline a separated function.

Reply:
done